### PR TITLE
Introduce computeGasData

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-commons/lib/core/utils/hexStrings.ts [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import {ensure} from './errors';
 
+export const isProperHexString = (hexString: string) => {
+  return hexString.startsWith('0x') && hexString.length % 2 === 0;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/hexStrings.ts]: Doesn't check the conent of string, simpler:
  return !!address.match(/^0x[0-9A-Fa-f]+$/);

File: universal-login-commons/test/core/utils/hexStrings.ts [hunk 2]
```diff
@@ -26,6 +26,36 @@ describe('UNIT: hex strings', () => {
...
+      expect(isProperHexString('0xFFFFFFe7d45c34110B34Ed269AD86248884E78C7BEEF')).to.be.true;
+    });
+
+    it('0x', () => {
+      expect(isProperHexString('0x')).to.be.true;
+    });
+
+    it('empty', () => {
+      expect(isProperHexString('')).to.be.false;
+    });
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/hexStrings.ts]: Missing test `0xINVALIDHEX`

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import {ensure} from '../errors';
+import {isProperHexString} from '../hexStrings';
+
+export const ZERO_BYTE_GAS_COST = 4;
+export const NON_ZERO_BYTE_GAS_COST = 68;
+
+export const computeGasData = (data: string) => {
+  ensure(isProperHexString(data), Error, 'Not a valid hex string');
+
+  if (data === '0x') {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/messages/computeGasData.ts]: duplicated check

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    return 0;
+  }
+
+  return data
+    .slice(2)
+    .match(/.{2}/g)!
+    .reduce((totalGasCost, byte) => {
+      const byteCost = byte === '00' ? ZERO_BYTE_GAS_COST : NON_ZERO_BYTE_GAS_COST;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/messages/computeGasData.ts]: extract function

File: universal-login-commons/test/core/utils/computeGasData.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  it('0x', async () => {
+    const data = '0x';
+
+    expect(computeGasData(data)).to.equal(0);
+  });
+
+  it('0xbeef', async () => {
+    const data = '0xbeef';
+
+    expect(computeGasData(data)).to.equal(2 * NON_ZERO_BYTE_GAS_COST);
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/computeGasData.ts]: put exact values

File: universal-login-sdk/lib/api/sdk.ts [hunk 6]
```diff
@@ -125,10 +125,12 @@ class UniversalLoginSDK {
...
   async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
     const unsignedMessage = {
-      ...this.sdkConfig.paymentOptions,
+      gasPrice: this.sdkConfig.paymentOptions.gasPrice,
+      gasToken: this.sdkConfig.paymentOptions.gasToken,
       ...message,
       nonce: message.nonce || parseInt(await this.getNonce(message.from!), 10),
-      gasData: utils.bigNumberify(0)
+      gasData: utils.bigNumberify(0),
+      gasLimitExecution: message.gasLimit || this.sdkConfig.paymentOptions.gasLimit,
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]:
```ts
gasLimit = message.gasLimit || this.sdkConfig.paymentOptions.gasLimit;
gasData = calcGasData(data);
...
gasData,
gasLimitExecution: gasLimit - gasData
```

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  ensure(isProperHexString(data), Error, 'Not a valid hex string');
+
+  if (data === '0x') {
+    return 0;
+  }
+
+  return data
+    .slice(2)
+    .match(/.{2}/g)!
+    .reduce((totalGasCost, byte) =>  totalGasCost + gasCostFor(byte), 0);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/messages/computeGasData.ts]: btw if you reverse match with slice you will not need extra if at the begining.
So with:
```ts
  return data
    .match(/.{2}/g)
    .slice(1)
    .reduce((totalGasCost, byte) =>  totalGasCost + gasCostFor(byte), 0);
```
you will not need preceding:
```ts
  if (data === '0x') {
    return 0;
  }
```

File: universal-login-sdk/lib/api/sdk.ts [hunk 8]
```diff
@@ -124,12 +124,19 @@ class UniversalLoginSDK {
...
 
   async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
+    const gasLimit = utils.bigNumberify(message.gasLimit || this.sdkConfig.paymentOptions.gasLimit);
+    const gasData = computeGasData(message.data as string);
+    const gasLimitExecution = gasLimit.sub(gasData);
+
     const unsignedMessage = {
-      ...this.sdkConfig.paymentOptions,
+      gasPrice: this.sdkConfig.paymentOptions.gasPrice,
+      gasToken: this.sdkConfig.paymentOptions.gasToken,
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: `message.gasTokes || this.sdkConfig.paymentOptions.gasToken`

File: universal-login-sdk/lib/api/sdk.ts [hunk 9]
```diff
@@ -140,7 +147,7 @@ class UniversalLoginSDK {
       ...transactionDetails,
       to,
       from: to,
-      data
+      data: data || '0x'
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: should raise an exception if undefined

File: universal-login-sdk/lib/api/sdk.ts [hunk 10]
```diff
@@ -124,12 +124,19 @@ class UniversalLoginSDK {
   }
 
   async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
+    const gasLimit = utils.bigNumberify(message.gasLimit || this.sdkConfig.paymentOptions.gasLimit);
+    const gasData = computeGasData(message.data as string);
+    const gasLimitExecution = gasLimit.sub(gasData);
+
     const unsignedMessage = {
-      ...this.sdkConfig.paymentOptions,
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: add type to `unsignedMessage` #sanitycheck

File: universal-login-sdk/lib/api/sdk.ts [hunk 11]
```diff
@@ -124,12 +124,19 @@ class UniversalLoginSDK {
   }
 
   async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
+    const gasLimit = utils.bigNumberify(message.gasLimit || this.sdkConfig.paymentOptions.gasLimit);
+    const gasData = computeGasData(message.data as string);
+    const gasLimitExecution = gasLimit.sub(gasData);
+
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: Extract `messageToUnsignedMessage`

File: universal-login-sdk/lib/api/sdk.ts [hunk 12]
```diff
@@ -123,20 +123,38 @@ class UniversalLoginSDK {
     );
   }
 
+  removeGasLimit(message: Partial<Message>) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: `withoutGasLimit`

Comment (gasiortomasz) [reply universal-login-sdk/lib/api/sdk.ts]: this is actually obsolete. Will remove this

File: universal-login-sdk/lib/api/sdk.ts [hunk 13]
```diff
@@ -124,19 +124,32 @@ class UniversalLoginSDK {
   }
 
   async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
-    const unsignedMessage = {
-      ...this.sdkConfig.paymentOptions,
-      ...message,
-      nonce: message.nonce || parseInt(await this.getNonce(message.from!), 10),
-      gasData: utils.bigNumberify(0)
-    } as MessageWithFrom;
+    const unsignedMessage = await this.messageToUnsignedMessage(message);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]:
```ts
const {paymentOptions} = this.sdkConfig;
const nonce = parseInt(await this.getNonce(message.from!), 10);
await messageToUnsignedMessage({...paymentOptions, nonce, ...message});
```

File: universal-login-sdk/test/unit/core/messageTransform.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import {expect} from 'chai';
+import {utils} from 'ethers';
+import {Message, TEST_CONTRACT_ADDRESS, SignedMessage} from '@universal-login/commons';
+import {SdkConfigDefault} from '../../../lib/config/SdkConfigDefault';
+import {messageToUnsignedMessage} from '../../../lib/core/utils/utils';
+
+describe('UNIT: messageToUnsignedMessage', () => {
+  it('correct transform', async () => {
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/unit/core/messageTransform.ts]: missing case: empty message

**Pull request merged**